File: PID.py
import csv
import time


class PID:
    def __init__(self, P=1, I=0, D=0, debug=True, file="PIDvars.csv"):
        self.P = P
        self.I = I
        self.D = D
        self.debug = debug
        self.iAccumulator = 0
        self.prevError = 0
        if self.debug:
            self.fileOutput = open(file, "w")
            self.writePointer = csv.writer(self.fileOutput)
            self.writePointer.writerow([
                "Equation", "I Accumulator", "Error", "Prev Error",
                "P With Error", "I with I Accumulator", "D with Prev Error",
                "Setpoint", "PIDInput", "Time"
            ])
        self.first = True

    def update(self, target, current):
        error = target - current
        self.iAccumulator += error
        if self.first:
            self.iAccumulator = 0
            self.prevError = error
            self.first = False
            self.sTime = time.time()
        output = (self.P * error) + (self.iAccumulator * self.I) + ((error - self.prevError) * self.D)
        if self.debug:
            self.writePointer.writerow([
                output,
                self.iAccumulator,
                error,
                self.prevError,
                self.P * error,
                self.I * self.iAccumulator,
                self.D * (error-self.prevError),
                target,
                current,
                time.time() - self.sTime
            ])
        self.prevError = error

        return output

    def reset(self):
        self.first = True

    def close(self):
        if self.debug:
            self.fileOutput.close()

File: CPP_Libraries/PID_CPP/PID.cpp
#include "PID.hpp"
#include <iostream>
#include <fstream>
#include <time.h>


PID::PID (double P_value, double I_value, double D_value, bool debug_value, std::string file_path) {
    P = P_value;
    I = I_value;
    D = D_value;
    debug = debug_value;
    first = true;
    iAccumulator = 0;
    prevError = 0;
    file.open(file_path);
    file << "Equation,I Accumulator,Error,Prev Error,P With Error,I with I Accumulator,D with Prev Error,Setpoint,Time\n";
}

double PID::update(double target, double current) {
    double error = current - target;
    iAccumulator += error;
    if (first) {
        iAccumulator = 0;
        prevError = error;
        first = false;
        sTime = time(0);
    }
    double output = (P * error) + (iAccumulator * I) + ((error - prevError) * D);

    if (debug) {
        file << output << "," << iAccumulator << "," << error << "," << prevError << "," << P * error << "," << I * iAccumulator << "," << D * (error - prevError) << "," << target << "," << time(0) - sTime;
    }

    return output;
}

void PID::reset() {
    first = true;
}

void PID::close() {
    file.close();
}
